drivers: adc: fix issue and add syscall for adc_read_async()

tyrantlizard77 opened the pull request:
We should not be storing the sequence pointer, as
adc_read_async() returns immediately. The memory could
be heap allocated, or on a call stack. Make a copy of
it instead.

With this bug out of the way, we add system call handlers
for adc_read_async(), with associated necessary
changes to the test case to support this scenario.

Fixes: #15039
Fixes: #14084 

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: adc: make a copy of sequence struct

We should not be storing the sequence pointer, as
adc_read_async() returns immediately. The memory could
be heap allocated, or on a call stack. Make a copy of
it instead.

Fixes: #15039

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

aescolar commented:
@andrewboie I remove the "priority: medium" label from here, as PRs priorities are not looked at by anybody following our current way of working. The bug it links to has already a priority.

anangl commented on drivers/adc/adc_context.h:
```diff
@@ -181,13 +183,16 @@ static inline void adc_context_complete(struct adc_context *ctx, int status)
-	if (ctx->sequence->options) {
+	if (sequence->options) {
+		(void)memcpy(&ctx->options, sequence->options,
+			     sizeof(struct adc_sequence_options));
```
Why not just use `ctx->options = sequence->options;` and let the compiler decide how to handle the copying? Especially when the structure is not big, it is likely that this could be optimized.
Same story above, with `ctx->sequence`.

anangl commented on drivers/adc/adc_context.h:
```diff
@@ -205,16 +210,16 @@ static inline void adc_context_start_read(struct adc_context *ctx,
+	if (ctx->sequence.options) {
 		adc_sequence_callback callback =
-			ctx->sequence->options->callback;
+			ctx->options.callback;
```
After the change, this line no longer needs to be broken in two.

andrewboie replied:
done

anangl reviewed (commented):
[This comment](https://github.com/zephyrproject-rtos/zephyr/blob/645e95926f5b0ed1a4015698486368483c93b69e/include/adc.h#L349) should be removed as well, as it becomes pointless when the sequence structure is copied.

Commit:
adc: add syscall for adc_read_async()

Same restriction as adc_read(), callback must be NULL.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
adc: doc: clarify callback pointer from user mode

They are forbidden since they run in supervisor mode.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: adc_api: run adc_read_async() from user

We define a system heap and assign our resource pool
from it as k_poll() requires an implicit allocation.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
adc: amend adc_read_async comment

This is no longer necessary since the sequence is copied.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andyross reviewed (approved):
Makes sense

anangl commented on drivers/adc/adc_context.h:
```diff
@@ -10,6 +10,7 @@
 
 #include <adc.h>
 #include <atomic.h>
+#include <string.h>
```
This is no longer needed.
